src/engine/openmp.cc

```diff
@@ -73,14 +73,18 @@ void OpenMP::set_reserve_cores(int cores) {
   CHECK_GE(cores, 0);
   reserve_cores_ = cores;
 #ifdef _OPENMP
-  omp_thread_max_ = std::max(omp_thread_max_ - reserve_cores_, 1);
+  if (reserve_cores_ >= omp_thread_max_) {
+    omp_set_num_threads(1);
+  } else {
+    omp_set_num_threads(omp_thread_max_ - reserve_cores_);
```

Comment:
would this still causing problem if this function called multiple times?

Reply:
this shouldnt cause a problem since omp_thread_max_ doesnt change. For example, for a 16 core machine, and eight GPU workers , the num threads for all of them will be set to 14.